rts/lra/lra-coordinator-jar/src/test/java/io/narayana/lra/coordinator/TestBase.java

```diff
@@ -0,0 +1,207 @@
+    @BeforeClass
+    public static void beforeClass() {
+        String host = System.getProperty(NarayanaLRAClient.LRA_COORDINATOR_HOST_KEY, "localhost");
+        String port = System.getProperty(NarayanaLRAClient.LRA_COORDINATOR_PORT_KEY, "8080");
+        String path = System.getProperty(NarayanaLRAClient.LRA_COORDINATOR_PATH_KEY, COORDINATOR_PATH_NAME);
+
+        coordinatorUrl = String.format("http://%s:%s/%s", host, port, path);
+        recoveryUrl = String.format("http://%s:%s/%s/%s/%s",
+                host, port, COORDINATOR_DEPLOYMENT, RECOVERY_COORDINATOR_PATH_NAME, RECOVERY_COORDINATOR_SUB_RESOURCE_NAME);
+        storeDir = Paths.get(String.format("%s/standalone/data/tx-object-store", System.getProperty("env.JBOSS_HOME", "null")));
+        jvmArgs = System.getProperty("server.jvm.args");
+
+        deploymentURL = String.format("http://%s:%s/%s", host, port, COORDINATOR_DEPLOYMENT);
+    }
+
+    @Before
+    public void before() throws URISyntaxException, MalformedURLException {
+        LRALogger.logger.infof("Starting test %s", testName);
+
+        lraClient = new NarayanaLRAClient(new URI(coordinatorUrl));
+    }
+
+    @After
+    public void after() {
+        LRALogger.logger.infof("Finised test %s", testName);
+        lraClient.close();
+    }
+
+    void startContainer(String bytemanScript) {
+        Config config = new Config();
+        String javaVmArguments = System.getProperty("server.jvm.args");
+
+        clearRecoveryLog();
+
+        if (bytemanScript != null) {
+            javaVmArguments = javaVmArguments.replaceAll("=listen", "=script:scripts/@BMScript@.btm,listen");
+            javaVmArguments = javaVmArguments.replace("@BMScript@", bytemanScript);
+        }
+
+        config.add("javaVmArguments", javaVmArguments);
+
+        containerController.start(COORDINATOR_CONTAINER, config.map());
+        deployer.deploy(COORDINATOR_DEPLOYMENT);
+    }
+
+    void restartContainer() {
+        try {
+            // ensure that the controller is not running
+            containerController.kill(COORDINATOR_CONTAINER);
+            System.out.printf("jboss-as kill worked%n");
+        } catch (Exception e) {
+            System.out.printf("jboss-as kill: %s", e.getMessage());
+        }
+
+        Config config = new Config();
+        String javaVmArguments = System.getProperty("server.jvm.args");
+        config.add("javaVmArguments", javaVmArguments);
+        containerController.start(COORDINATOR_CONTAINER);
+    }
+
+    void stopContainer() {
+        if (containerController.isStarted(COORDINATOR_CONTAINER)) {
+            LRALogger.logger.info("STOPPING CONTAINER%n");
+
+            deployer.undeploy(COORDINATOR_DEPLOYMENT);
+
+            containerController.stop(COORDINATOR_CONTAINER);
+            containerController.kill(COORDINATOR_CONTAINER);
+        }
+    }
+
+    int recover() throws URISyntaxException {
+        try (Response response = ClientBuilder.newClient().target(new URI(recoveryUrl))
```

Comment:
Is the created client also closed in try-with-resources?

Reply:
I am not even sure if [Client](https://docs.oracle.com/javaee/7/api/javax/ws/rs/client/Client.html) is `AutoCloseable`. If it was I would guess it could need something like the following (assuming it would compile) to work:
```try (Client client = ClientBuilder.newClient(); Response resource = client.target(...```

Reply:
but I still close it usually when I type rest client invocations. But you are right it's not AutoClosable. Not sure if this it is necessary to close it.

Reply:
Agreed, I was not suggesting not to close it - just trying to develop the thread. If it's not autocloseable then I think we just need the try-with-resources wrapped with a try/finally block

Reply:
The call that closes the underlying connections/resources is either when the response is closed or when the response entity is fully read (which is what the above code does in its `try with resources code block`).

The spec has this to say about the JAX-RS client API close method:

> Calling this method effectively invalidates all resource targets produced by the client instance. Invoking any method on such targets once the client is closed would result in an IllegalStateException being thrown.

So since I am not using any of those targets after the call finishes (because the client is implicitly created) I don't need to call close on it.

Reply:
How about this statement "Client instances must be properly closed before being disposed to avoid leaking resources." from https://docs.oracle.com/javaee/7/api/javax/ws/rs/client/Client.html

Reply:
Maybe later on in the project we can do some resource utilisation testing and performance optimisations but let's do it as part of a more comprehensive analysis of the non functional behaviour of the system.

Reply:
Okay, it is just a test but fair enough. If that is the only pending issue then I will push a fix for this comment?